sklearn/preprocessing/tests/test_encoders.py

```diff
@@ -404,55 +404,78 @@ def test_one_hot_encoder_inverse():
         assert_raises_regex(ValueError, msg, enc.inverse_transform, X_tr)
 
 
-def test_one_hot_encoder_categories():
-    X = [['abc', 1, 55], ['def', 2, 55]]
-
+@pytest.mark.parametrize("X, cat_exp, cat_dtype", [
+    ([['abc', 55], ['def', 55]], [['abc', 'def'], [55]], np.object_),
+    (np.array([[1, 2], [3, 2]]), [[1, 3], [2]], np.integer),
+    (np.array([['A', 'cat'], ['B', 'cat']], dtype=object),
+     [['A', 'B'], ['cat']], np.object_),
+    (np.array([['A', 'cat'], ['B', 'cat']]),
+     [['A', 'B'], ['cat']], np.str_)
+    ], ids=['mixed', 'numeric', 'object', 'string'])
+def test_one_hot_encoder_categories(X, cat_exp, cat_dtype):
     # order of categories should not depend on order of samples
     for Xi in [X, X[::-1]]:
-        enc = OneHotEncoder()
+        enc = OneHotEncoder(categories='auto')
         enc.fit(Xi)
         # assert enc.categories == 'auto'
         assert isinstance(enc.categories_, list)
-        cat_exp = [['abc', 'def'], [1, 2], [55]]
         for res, exp in zip(enc.categories_, cat_exp):
             assert res.tolist() == exp
-
-
-def test_one_hot_encoder_specified_categories():
-    X = np.array([['a', 'b']], dtype=object).T
-
-    enc = OneHotEncoder(categories=[['a', 'b', 'c']])
+            assert np.issubdtype(res.dtype, cat_dtype)
+
+
+@pytest.mark.parametrize("X, X2, cats, cat_dtype", [
+    (np.array([['a', 'b']], dtype=object).T,
+     np.array([['a', 'd']], dtype=object).T,
+     [['a', 'b', 'c']], np.object_),
+    (np.array([[1, 2]], dtype='int64').T,
+     np.array([[1, 4]], dtype='int64').T,
+     [[1, 2, 3]], np.int64),
+    (np.array([['a', 'b']], dtype=object).T,
```

Comment:
I can't see the difference between this and the first case above...

Reply:
In the first case the categories are specified as a list, in this case (two lines below), as an array of string dtype (to test that this is converted to array of object dtype). 
In the `ids` I tried to make this clear with the name.